electrum/lnrouter.py

```diff
@@ -161,12 +167,251 @@ def is_fee_sane(fee_msat: int, *, payment_amount_msat: int) -> bool:
+        hint_key = self._hint_key(node_from, node_to, channel_id)
+        hint = self._liquidity_hints.get(hint_key)
+        if not hint:
+            hint = self._liquidity_hints[hint_key] = LiquidityHint()
+        hint.update_cannot_send(node_from < node_to, amount)
+
+    def penalty(self, node_from: bytes, node_to: bytes, channel_id: ShortChannelID, amount: int) -> float:
+        """Gives a penalty when sending from node1 to node2 over channel_id with an
+        amount in units of millisatoshi.
+
+        The penalty depends on the can_send and cannot_send values that was
+        possibly recorded in previous payment attempts.
+
+        A channel that can send an amount is assigned a penalty of zero, a
+        channel that cannot send an amount is assigned an infinite penalty.
+        If the sending amount lies between can_send and cannot_send, there's
+        uncertainty and we give a default penalty. The default penalty
+        serves the function of giving a positive offset (the Dijkstra
+        algorithm doesn't work with negative weights), from which we can discount
+        from. There is a competition between low-fee channels and channels where
+        we know with some certainty that they can support a payment. The penalty
+        ultimately boils down to: how much more fees do we want to pay for
+        certainty of payment success? This can be tuned via DEFAULT_PENALTY_BASE_MSAT
+        and DEFAULT_PENALTY_PROPORTIONAL_MILLIONTH. A base _and_ relative penalty
+        was chosen such that the penalty will be able to compete with the regular
+        base and relative fees.
+        """
+        hint = self.get_hint(node_from, node_to, channel_id)
+        can_send = hint.can_send(node_from < node_to)
+        cannot_send = hint.cannot_send(node_from < node_to)
+
+        # if we know nothing about the channel, return a default penalty
+        if (can_send, cannot_send) == (None, None):
+            return fee_for_edge_msat(amount, DEFAULT_PENALTY_BASE_MSAT, DEFAULT_PENALTY_PROPORTIONAL_MILLIONTH)
+        # next cases are with half information
+        elif can_send and not cannot_send:
+            if amount <= can_send:
+                return 0
+            else:
+                return fee_for_edge_msat(amount, DEFAULT_PENALTY_BASE_MSAT, DEFAULT_PENALTY_PROPORTIONAL_MILLIONTH)
+        elif not can_send and cannot_send:
+            if amount >= cannot_send:
+                return inf
+            else:
+                return fee_for_edge_msat(amount, DEFAULT_PENALTY_BASE_MSAT, DEFAULT_PENALTY_PROPORTIONAL_MILLIONTH)
+        # we know how much we can/cannot send
+        elif can_send and cannot_send:
+            if amount <= can_send:
+                return 0
+            elif amount < cannot_send:
+                return fee_for_edge_msat(amount, DEFAULT_PENALTY_BASE_MSAT, DEFAULT_PENALTY_PROPORTIONAL_MILLIONTH)
+            else:
+                return inf
+        return 0
```

Comment:
```suggestion
        if cannot_send is not None and amount >= cannot_send:
            return inf
        if can_send is not None and amount <= can_send:
            return 0
        return fee_for_edge_msat(amount, DEFAULT_PENALTY_BASE_MSAT, DEFAULT_PENALTY_PROPORTIONAL_MILLIONTH)
```

Reply:
Thank you for this, a nice simplification!